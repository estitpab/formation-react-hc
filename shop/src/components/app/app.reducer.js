/* version "immutable.js":
export const initialState = Immutable.fromJS({
  categories: {
    loading: false,
    error: null,
    result: null
  },
  currentCategory: null
});
*/

export const initialState = {
  categories: {
    loading: false,
    error: null,
    result: null // Array<string>
  },
  currentCategory: null
};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case "GET_CATEGORIES_START":
      // Immutable:
      // return state
      //   .set('currentCategory', null)
      //   .set('categories', Map({ ... }))
      return {
        ...state,
        categories: { loading: true, error: null, result: null },
        currentCategory: null
      };

    case "GET_CATEGORIES_SUCCESS":
      return {
        ...state,
        categories: {
          loading: false,
          error: null,
          result: action.payload.result
        },
        currentCategory: action.payload.result[0]
      };

    case "GET_CATEGORIES_ERROR":
      return {
        ...state,
        categories: { loading: false, error: action.error, result: null }
      };

    case "SET_CATEGORY":
      return {
        ...state,
        currentCategory: action.payload.category
      };

    default:
      return state;
  }
};
