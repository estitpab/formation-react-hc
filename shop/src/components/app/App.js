import React from "react";

import "./App.scss";

import Nav from "./Nav";
import ProductList from "../products/ProductList";
import CartModal from "../cart/CartModal";

function App() {
  //const [value, setValue] = React.useState("truc");
  return (
    <div className="App">
      {/*
      <form
        style={{ margin: "1em", padding: "1em", border: " 1px solid black" }}
        onSubmit={e => {
          e.preventDefault();
          console.log(e.currentTarget.elements.input1.value);
        }}
      >
        Input non contrôlé :
        <input name="input1" defaultValue="truc" />
        Input contrôlé (readOnly) :
        <input value="truc" readOnly />
        Input contrôlé (onChange) :
        <input
          value={value}
          onChange={e => setValue(e.target.value)}
          required
          pattern="truc|bidule"
        />
        <input type="submit" />
      </form>
      <hr />
      */}
      <Nav />
      <main>
        <div className="container">
          <ProductList />
        </div>
        <CartModal />
      </main>
    </div>
  );
}

export default App;
