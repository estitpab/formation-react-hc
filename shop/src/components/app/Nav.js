import React from "react";
import CartButton from "../cart/CartButton";
import { useSelector, useDispatch } from "react-redux";
import { setCurrentCategory } from "./app.actions";
import cx from "classnames";
import { computeCartCount } from "../cart/cart.utils";

const titlize = string =>
  string[0].toUpperCase() + string.substring(1).toLowerCase() + "s";

/*
class Nav extends React.Component {
  state = { categories: [], loading: true, error }
  componentDidMount() {
    getCategories()
      .then(categories => this.setState({ loading: false, categories }))
      .catch(error => this.setState({ loading: false, error }))
  }
  render() {
    const { categories, loading, error } = this.state
    ...
  }
}
*/

/*
const mapStateToProps = state => ({
  loading: state.categories.loading,
  categories: state.categories.result,
  error: state.categories.error
})

const Nav = ({ loading, categories, error, count }) => {
  ...
}

export default connect(mapStateToProps)(Nav) // ({ count }) => ...
*/

const Nav = () => {
  // useSelector:
  // 1. useContext() pour aller chercher le store
  // 2. useState() pour pouvoir faire évoluer le composant (valeur initiale = calculée depuis la fonction de sélection)
  // 3. store.subscribe pour mettre à jour le state

  /*
  const { loading, categories, error } = useSelector(state => ({
    loading: state.categories.loading,
    categories: state.categories.result,
    error: state.categories.error
  }));
  */

  const loading = useSelector(state => state.app.categories.loading);
  const categories = useSelector(state => state.app.categories.result);
  const error = useSelector(state => state.app.categories.error);
  const currentCategory = useSelector(state => state.app.currentCategory);
  // const cartCount = useSelector(state => computeCartCount(state.cart.cart.result));
  const cartCount = useSelector(state => state.cart.cartCount);

  const dispatch = useDispatch();

  return (
    <nav className="Nav">
      <div className="nav-wrapper">
        <a href="/" className="brand-logo">
          Bazket
        </a>
        <ul id="nav-mobile" className="right hide-on-med-and-down">
          {categories &&
            categories.map(category => (
              <li
                key={category}
                className={cx({ active: category === currentCategory })}
              >
                <a
                  href={`/category/${category}`}
                  onClick={e => {
                    e.preventDefault();
                    dispatch(setCurrentCategory(category));
                  }}
                >
                  {titlize(category)}
                </a>
              </li>
            ))}
          {loading && null /* TODO spinner? */}
          {error && (
            // TODO better error?
            <li>
              <strong>{error.message}</strong>
            </li>
          )}
          <li>
            <CartButton count={cartCount} />
          </li>
        </ul>
      </div>
    </nav>
  );
};

export default Nav;
