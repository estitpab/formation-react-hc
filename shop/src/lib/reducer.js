import { combineReducers } from "redux";
import * as app from "../components/app/app.reducer";
import * as cart from "../components/cart/cart.reducer";
// import * as products from '../components/products/products.reducer';

export const initialState = {
  app: app.initialState,
  cart: cart.initialState
};

export const reducer = combineReducers({
  app: app.reducer,
  cart: cart.reducer
});

/*
// List modules here: initialState and rootReducer will be computed from this list
const modules = {
  app,
  cart,
  //products,
};

// Compute exported initialState from reducers' initial states
export const initialState = Object.keys(modules).reduce((state, key) => {
  state[key] = modules[key].initialState; // eslint-disable-line no-param-reassign
  return state;
}, {});

// Combine small reducers to generate the global state and reducer
export const reducer = combineReducers(
  Object.keys(modules).reduce((reducers, key) => {
    reducers[key] = modules[key].reducer; // eslint-disable-line no-param-reassign
    return reducers;
  }, {}),
);
*/
