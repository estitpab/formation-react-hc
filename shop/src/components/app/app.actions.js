import { getCategories } from "../../lib/api";

export const fetchCategories = () => dispatch => {
  dispatch({
    type: "GET_CATEGORIES_START"
  });
  getCategories()
    .then(cart =>
      dispatch({
        type: "GET_CATEGORIES_SUCCESS",
        payload: { result: cart }
      })
    )
    .catch(error =>
      dispatch({
        type: "GET_CATEGORIES_ERROR",
        error
      })
    );
};

export const setCurrentCategory = category => ({
  type: "SET_CATEGORY",
  payload: { category: category }
});
