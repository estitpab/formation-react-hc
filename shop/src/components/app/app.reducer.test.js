import { initialState, reducer } from "./app.reducer";
import {
  getCategoriesStart,
  getCategoriesSuccess,
  getCategoriesError
} from "./app.actions";

// TODO fix (on a virÃ© les action creators)
describe.skip("App reducer", () => {
  test("start", () => {
    const state = reducer(initialState, getCategoriesStart());
    expect(state.categories).toHaveProperty("loading", true);
  });

  test("success", () => {
    const state = reducer(initialState, getCategoriesSuccess(["toto"]));
    expect(state.categories).toHaveProperty("loading", false);
    expect(state.categories).toHaveProperty("result", ["toto"]);
  });

  test("error", () => {
    const state = reducer(initialState, getCategoriesError(new Error("oops")));
    expect(state).toMatchSnapshot();
  });
});
