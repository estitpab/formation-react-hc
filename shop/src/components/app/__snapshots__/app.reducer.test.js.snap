// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App reducer error 1`] = `
Object {
  "categories": Object {
    "error": [Error: oops],
    "loading": false,
    "result": null,
  },
  "currentCategory": null,
}
`;
